RP2: Support for DMA Timers

nickovs opened the pull request:
### Summary

The RP2 DMA controller includes four pacing timers that can be used to control the rate of transfers. This PR adds a new `DMATimer` class to make it easy to use these timers with DMA channels being accessed through the existing `DMA` class.

The new `DMATimer` class allows the user to claim a specific timer or for one to be selected automatically from the free timers. Once selected, the rate can be set either by directly setting the divider `ratio` attribute to an X,Y tuple  or to a desired frequency by setting the `freq` attribute. In the later case the divider ratio will silently be set to the value that yields the closest frequency and the actual frequency (and divider ratio) can be read back using the attributes. The correct value needed for the `treq_sel` in the DMA configuration can be read from an attribute of the same name.

### Testing

The code has been tested on Pi Pico W and Pi Pico2 W boards, but should work directly on any board that supports the RP2 DMA class.

Setting the divider ratio, both directly and indirectly by setting the frequency, has been tested both for checks on input value ranges and also to ensure that the DMA transfer rates are correctly throttled by (performing timing checks for memory-memory transfers at various rates).

### Trade-offs and Alternatives

While this code is fairly small, a substantial part of the code is the Stern-Brocot search for the best divider ratio for the target frequency. This code is already simplified to remove the more complex check for the true best fit semi-convergent, accepting a good fit in exchange for removing some complex code, but it could be made even simpler is less good fits were acceptable. Note that unlike the divider for the SPI controller, the DMA timers support 16 bit numerator and denominator for the divider, so the linear searches used there are not appropriate here since the search space is too large for that to be efficient.

nickovs committed:
rp2/rp2_dma: Add support for DMA pacing timers.

Signed-off-by: Nicko van Someren <nicko@nicko.org>

dpgeorge commented on ports/rp2/rp2_dma.c:
```diff
@@ -56,6 +57,194 @@ typedef struct _rp2_dma_ctrl_field_t {
+
+static void rp2_dma_timer_print(const mp_print_t *print, mp_obj_t self_in, mp_print_kind_t kind) {
+    rp2_dma_timer_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    mp_printf(print, "DMATimer(%u)", self->timer_id);
```
Might be smaller code to do `mp_printf(print, "%q(%u)", MP_QSTR_DMATimer, self->timer_id);`

nickovs replied:
I confess that I wasn't aware of `%q` in `mp_printf()`! Yes, that would save some bytes.

dpgeorge commented on ports/rp2/rp2_dma.c:
```diff
@@ -56,6 +57,194 @@ typedef struct _rp2_dma_ctrl_field_t {
+    // Find the values x and y such that x/y is the closest faction to p/q for which
+    // both terms are less than DMA_TIMER_MAX_TERMS.
+
+    // This implementation comptes the continued fraction and then checks if the
```
"computes"

dpgeorge commented on ports/rp2/rp2_dma.c:
```diff
@@ -56,6 +57,194 @@ typedef struct _rp2_dma_ctrl_field_t {
+        uint32_t k_next = a * k_curr + k_prev;
+
+        if (h_next >= DMA_TIMER_MAX_TERMS || k_next >= DMA_TIMER_MAX_TERMS) {
+            // The next convergent would overflow. Check of there is obviously a better semi-convergent.
```
of -> if

dpgeorge commented on ports/rp2/rp2_dma.c:
```diff
@@ -56,6 +57,194 @@ typedef struct _rp2_dma_ctrl_field_t {
+
+        if (h_next >= DMA_TIMER_MAX_TERMS || k_next >= DMA_TIMER_MAX_TERMS) {
+            // The next convergent would overflow. Check of there is obviously a better semi-convergent.
+            // If there is a value os n that is >= a/2 then it will always be better than the current ratio.
```
os -> of

nickovs replied:
I left the typos in so that you'd know that the PR wasn't written by AI ;-)

dpgeorge replied:
Haha!  Excellent.  The thought did cross my mind.

And you now also know that I did a thorough review :)

dpgeorge commented on docs/library/rp2.DMA.rst:
```diff
@@ -291,3 +295,55 @@ below is a Pythonic version of the example in sub-section 2.5.6.2.
+
+.. attribute:: DMATimer.treq_sel
+
+    Read the value that needs to be passed to `DMA.pack_ctrl` in order to use this timer
```
suggest "... that needs to be passed to the `treg_sel` argument of `DMA.pack_ctrl` ..." to make it a bit clearer how the value is used

nickovs replied:
Noted. Or I could implement MP_UNARY_OP_INT to return the selector index so that you can just pass if directly to the `treq_sel` of the config as you suggested above.

dpgeorge replied:
Ah, yes, that's quite a good idea, to implement MP_UNARY_OP_INT.  That would allow the user to still retrieve the value if needed, and also mean zero changes to `pack_ctrl()`.

dpgeorge commented on docs/library/rp2.DMA.rst:
```diff
@@ -14,6 +14,10 @@ RP2040's DMA controller has 12 independent DMA channels that can run concurrentl
 
+The companion class :class:`DMATimer` provides access to the DMA controller's pacing timers.
+These timers can be used to control the speed of transfers into memory or preipherals that do
+not have their one transfer request signalling.
```
one -> own

dpgeorge commented on ports/rp2/rp2_dma.c:
```diff
@@ -56,6 +57,194 @@ typedef struct _rp2_dma_ctrl_field_t {
+            // Value needs to be a 2-tuple
+            size_t count = 2;
+            mp_obj_t *num_dom;
+            mp_obj_tuple_get(dest[1], &count, &num_dom);
```
suggest using `mp_obj_get_array_fixed_n()` instead, which checks the size and allows either a tuple or list

dpgeorge commented on ports/rp2/rp2_dma.c:
```diff
@@ -56,6 +57,194 @@ typedef struct _rp2_dma_ctrl_field_t {
+            mp_int_t numerator = mp_obj_get_int(num_dom[0]);
+            mp_int_t denominator = mp_obj_get_int(num_dom[1]);
+            if (numerator < 1 || numerator > 65535 || denominator < 1 || denominator > 65535 || numerator > denominator) {
+                mp_raise_ValueError(MP_ERROR_TEXT("value out of range"));
```
suggest a more specific error message like "ratio out of range"

nickovs replied:
I was going to do that, but `"value out of range"` is already in the ROM so this message doesn't cost anything. I'm happy either way.

dpgeorge replied:
IMO "value out of range" is not much more information than just a blank error message, ie just `mp_raise_ValueError(NULL)`.

In `ports/rp2/machine_pwm.c` we do have specific error messages "freq too large" and "freq too small".  Maybe it's worth reusing those messages here, and distinguishing both those cases?

nickovs replied:
Actually, the test here is checking when the user is trying to directly set the pair values for the divider, so it's not about the frequency per se. For the message to be more informative we'd need to be able to deliver four different messages, "[numerator/denominator] is too [small/large]". 

I'm inclined to thing that "value out of range" usefully distinguishes this class of error from the sorts of ValueError that might come from `mp_obj_get_int()`, and the message already exists so it's low cost, but let me know if you have an opinion either way.

dpgeorge replied:
> Actually, the test here is checking when the user is trying to directly set the pair values for the divider,

Ah, yes.  The frequency check is the error message above.

Happy to keep this as you have it.

dpgeorge reviewed (commented):
This looks like a good addition thank you.  A nice way to expose the hardware capabilities.

I kind of expected to see `freq` and `ratio` keyword arguments to the `DMATimer()` constructor.  But what you've implemented is the most minimal thing, which is good (because you'd probably want to update the freq/ratio on the same object, so setting that attribute is necessary to have, and then you don't need to be able to set it in the constructor).

But instead of exposing the `treq_sel` attribute, what do you think about being able to directly pass in a `DMATimer` instance to `DMA.pack_ctrl`?  Something like:
```py
dma_timer = DMATimer()
dma_timer.freq = F
dma.pack_ctrl(size=N, treq_sel=dma_timer)
```
That doesn't save much typing, but could be more readable, maybe?

I was also thinking if having an object is necessary, maybe the whole thing could be implemented in a (static class) function like `DMA.config_timer(pacing_timer_id, freq)` which would return the `treq_sel` value.  Eg:
```py
dma.pack_ctrl(size=N, treq_sel=dma.config_timer(1, F))
```
That's quite simple, but then you manually need to manage the pacing timer id, which is not ideal.

(I didn't test this PR, that's something I will do.)

dpgeorge commented:
> While this code is fairly small, a substantial part of the code is the Stern-Brocot search for the best divider ratio for the target frequency.

I did not know about this algorithm!  In the future we may be able to factor that out into a general helper function that can be used in other places and other ports.

nickovs commented:
@dpgeorge I think that this commit addresses all of the code and documentation issues you raised.

I've also updated the `DMATimer` constructor to accept `ratio` and `freq` keyword (only) arguments so you can grab a timer and set its frequency in a single call. Note that using the `dma.pack_ctrl(treq_sel=DMATimer(freq=44100))` pattern is a little risky since the lock on the underlying timer will get released when the `DMATimer` object gets garbage-collected, so there's a chance that some other code will ask for a timer, get the same one and change its frequency. That said, being able the assign the timer and set it up in a single call still makes sense and I've added a note in the documentation about this gotcha.

Let me know if there are any other changes you'd like.

dpgeorge commented on docs/library/rp2.DMA.rst:
```diff
@@ -291,3 +296,74 @@ below is a Pythonic version of the example in sub-section 2.5.6.2.
+where ``X < Y``.
+
+:class:`DMATimer` objects can be used directly as the value for treq_sel passed into the
+:meth:`DMA.unpack_ctrl()` function, since the value if ``int(dma_timer)`` is the index
```
Should this be `DMA.pack_ctrl()`?

nickovs replied:
Yes! Sorry about that.

dpgeorge commented on docs/library/rp2.DMA.rst:
```diff
@@ -291,3 +296,74 @@ below is a Pythonic version of the example in sub-section 2.5.6.2.
+    selected frequency, to the nearest 1Hz. The requested value needs be less than or equal to the
+    system clock speed and greater than or equal to 1/65535 of the system clock.
+
+.. attribute:: DMATimer.treq_sel
```
I think this attributed should be removed, now that it's possible to do the same thing via `int(dma_timer)`.  That saves code size, and means there's only one way to do the thing (keeping in spirit with the zen of Python).

nickovs replied:
That makes sense. It also makes it so that all of the attributes are read/write, as opposed to having one odd read-only attribute.

dpgeorge commented on ports/rp2/rp2_dma.c:
```diff
@@ -50,14 +51,225 @@ typedef struct _rp2_dma_obj_t {
+
+    if (dest[0] == MP_OBJ_NULL) {
+        // Load attribute
+        if (attr_in == MP_QSTR_treq_sel) {  // Use the same name (and QSTR) as the CTRL register field
```
As above, I think this attribute accessor should be removed.

dpgeorge reviewed (commented):
Thanks for updating.  The addition of `int(dma_timer)` is really neat.

Just a few additional comments.

Also, do you have time to add a test?  See `tests/ports/rp2/rp2_dma.py` for some inspiration that uses unittest (although I'd suggest making a new test `tests/ports/rp2/rp2_dma_timer.py`).  If you don't have time let me know and I'll add something.

dpgeorge commented:
One more thing: could you please rebase this on latest master so the CI can run?  There's a minor conflict due to the `const` added in cf0ef5a9d76208406707de64e83c16569c167513

nickovs commented:
I've removed the superfluous `treq_sel` attribute, updated the documentation and added some tests.

dpgeorge commented:
Please check the CI, eg `undefined reference to 'rp2_dma_timer_type'`.

nickovs commented:
I've rebased this to the current master and fixed and the CI issues. Hopefully this is good to go.

dpgeorge commented:
Thanks for updating.  I ran the tests on RPI_PICO, RPI_PICO2 and RPI_PICO2 in RISCV mode (`make BOARD=RPI_PICO2 BOARD_VARIANT=RISCV`).  RPI_PICO2 passes, but the other two (rp2040 and rp2350 in RISC-V mode) fail the speed throttle test, eg:

```
test_printing (__main__.Test) ... ok
test_set_ratio_unity (__main__.Test) ... ok
test_set_ratio (__main__.Test) ... ok
test_set_freq (__main__.Test) ... ok
test_speed_throttle (__main__.Test) ... FAIL

======================================================================
FAIL: test_speed_throttle <class 'Test'>
----------------------------------------------------------------------
Traceback (most recent call last):
  File "unittest/__init__.py", line 399, in run_one
  File "<stdin>", line 73, in test_speed_throttle
  File "unittest/__init__.py", line 187, in assertIn
AssertionError: Expected 75 to be in range(80, 120)

----------------------------------------------------------------------
Ran 5 tests

FAILED (failures=1, errors=0)
```

The "Expected 75 ..." goes down to about "Expected 68 ..." in some cases.

The failure is in the last part of the throttle test.  It seems that the slower MCUs have more overhead and so the very fast DMA transfer has a bit too much overhead in non-DMA activity.

Printing the time taken vs the speed I get:
```
# speed t_us
5000 51199
50000 5138
500000 533
5000000 71
```
The last line is too slow (71 us should be more like 53 us).

To fix it I suggest just doubling the amount of data, `dest = bytearray(2048)`.  That gets it working for me on all 3 boards (although RPI_PICO is just on the edge of the lower bound of `range(80, 120)`...).

dpgeorge commented on tests/cpydiff/modules_errno_enotsup.py:
```diff
@@ -1,10 +0,0 @@
-"""
```
I think this file was removed by mistake!

nickovs replied:
Sorry about that. I thought I had not checked that change in.

FWIW, I can't build the docs without taking that file out; with it there I get an error about there not being a difference when one is expected. This may be a result of me building on macOS rather than Linux but it's been a problem for a while.

nickovs replied:
I checked this. The test is:
```
print(f"{errno.errorcode[errno.EOPNOTSUPP]=!s}")
```
On Linux this prints:
```
errno.errorcode[errno.EOPNOTSUPP]=ENOTSUP
```
but on macOS (and possibly on other BSD-based Unix platforms) the output is:
```
errno.errorcode[errno.EOPNOTSUPP]=EOPNOTSUPP
```
That output is the same as you get from MicroPython, so when I try to build the docs I get a "no difference" error and it fails.

I'll raise a separate issue for this.

dpgeorge replied:
Might be worth building the docs in a macOS container in the github actions.  That should catch this problem.

nickovs replied:
Yes, that would work. Anyway, I restored the file in the last commit, so I think it should be all good now.

nickovs commented:
> Thanks for updating. I ran the tests on RPI_PICO, RPI_PICO2 and RPI_PICO2 in RISCV mode (`make BOARD=RPI_PICO2 BOARD_VARIANT=RISCV`). RPI_PICO2 passes, but the other two (rp2040 and rp2350 in RISC-V mode) fail the speed throttle test, eg:
...
> The "Expected 75 ..." goes down to about "Expected 68 ..." in some cases.
> 
> The failure is in the last part of the throttle test. It seems that the slower MCUs have more overhead and so the very fast DMA transfer has a bit too much overhead in non-DMA activity.
> 
> Printing the time taken vs the speed I get:
> 
> ```
> # speed t_us
> 5000 51199
> 50000 5138
> 500000 533
> 5000000 71
> ```
> 
> The last line is too slow (71 us should be more like 53 us).

Indeed. There's some overhead to the starting, stopping and capturing the times, and when the transfer time is small the overhead gets in the way.

> To fix it I suggest just doubling the amount of data, `dest = bytearray(2048)`. That gets it working for me on all 3 boards (although RPI_PICO is just on the edge of the lower bound of `range(80, 120)`...).

I can do that. Also, the ratios between differences of successive times is almost exactly 10 for all them. I will reenforce the test.

dpgeorge reviewed (approved):
Thanks for improving the test, it looks good now.  I tested on RPI_PICO_W, RPI_PICO2 and RPI_PICO2-RISCV.  The tests all pass on all boards.
